fvm/evm/emulator/emulator.go

```diff
@@ -0,0 +1,415 @@
+		config: cfg,
+		evm: gethVM.NewEVM(
+			*cfg.BlockContext,
+			*cfg.TxContext,
+			execState,
+			cfg.ChainConfig,
+			cfg.EVMConfig,
+		),
+		state: execState,
+	}, nil
+}
+
+// TODO errors returns here might not be fatal, what if we have hit the interaction limit ?
+func (bl *BlockView) commit(rootHash gethCommon.Hash) error {
+	// sets root hash
+	err := bl.database.SetRootHash(rootHash)
+	if err != nil {
+		return types.NewFatalError(err)
+	}
+
+	// commit atree changes back to the backend
+	err = bl.database.Commit()
+	if err != nil {
+		return types.NewFatalError(err)
+	}
+
+	return nil
+}
+
+type procedure struct {
+	config *Config
+	evm    *gethVM.EVM
+	state  *gethState.StateDB
+}
+
+// commit commits the changes to the state.
+// if error is returned is a fatal one.
+func (proc *procedure) commit() (gethCommon.Hash, error) {
+	// commits the changes from the journal into the in memory trie.
+	// in the future if we want to move this to the block level we could use finalize
+	// to get the root hash
+	newRoot, err := proc.state.Commit(true)
+	if err != nil {
+		return gethTypes.EmptyRootHash, types.NewFatalError(err)
+	}
+
+	// flush the trie to the lower level db
+	// the reason we have to do this, is the original database
+	// is designed to keep changes in memory until the state.Commit
+	// is called then the changes moves into the trie, but the trie
+	// would stay in memory for faster trasnaction execution. you
+	// have to explicitly ask the trie to commit to the underlying storage
+	err = proc.state.Database().TrieDB().Commit(newRoot, false)
+	if err != nil {
+		return gethTypes.EmptyRootHash, types.NewFatalError(err)
+	}
+	return newRoot, nil
+}
+
+func (proc *procedure) mintTo(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+	}
+
+	// create account if not exist
+	if !proc.state.Exist(addr) {
+		proc.state.CreateAccount(addr)
+	}
+
+	// add balance
+	proc.state.AddBalance(addr, amount)
+
+	// we don't need to increment any nonce, given the origin doesn't exist
+	res.StateRootHash, err = proc.commit()
+
+	return res, err
+}
+
+func (proc *procedure) withdrawFrom(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+	}
+
+	// check source balance
+	// if balance is lower than amount return
+	if proc.state.GetBalance(addr).Cmp(amount) == -1 {
+		return res, types.ErrInsufficientBalance
+	}
+
+	// sub balance
+	proc.state.SubBalance(addr, amount)
+
+	// we increment the nonce for source account cause
+	// withdraw counts as a transaction
+	nonce := proc.state.GetNonce(addr)
+	proc.state.SetNonce(addr, nonce+1)
+
+	res.StateRootHash, err = proc.commit()
+	return res, err
+}
+
+func (proc *procedure) run(msg *gethCore.Message) (*types.Result, error) {
```

Comment:
this method still needs improvements for error handling, but I'll don't want to hold on other PRs waiting, will create new PRs at the end.